File: impl/go/carriage/carriage_test.go
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.

package carriage_test

import (
	"bytes"
	"encoding/csv"
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"testing"

	"github.com/bubblefish-tech/naalp_protocol/impl/go/carriage"
	"github.com/bubblefish-tech/naalp_protocol/impl/go/cbor"
	"github.com/bubblefish-tech/naalp_protocol/impl/go/cose"
	"github.com/bubblefish-tech/naalp_protocol/impl/go/envelope"
	"github.com/cloudflare/circl/sign/mldsa/mldsa65"
)

type carriageCase struct {
	ProtocolID     uint64 `json:"protocol_id"`
	Class          uint64 `json:"class"`
	ContentType    uint64 `json:"content_type"`
	CorrelationHex string `json:"correlation_hex"`
	Method         string `json:"method"`
	ForeignHex     string `json:"foreign_hex"`
	BodyHex        string `json:"body_hex"`
}

var classDirs = []struct {
	Dir   string
	Class uint64
}{
	{"jsonrpc", carriage.ClassJSONRPC},
	{"http", carriage.ClassHTTP},
	{"msg", carriage.ClassMSG},
	{"stream", carriage.ClassSTREAM},
	{"doc", carriage.ClassDOC},
	{"opaque", carriage.ClassOPAQUE},
}

func loadClass(t *testing.T, dir string) carriageCase {
	t.Helper()
	b, err := os.ReadFile(filepath.Clean(filepath.Join("../../../vectors/carriage", dir, "cases.json")))
	if err != nil {
		t.Fatalf("read %s: %v", dir, err)
	}
	var c carriageCase
	if err := json.Unmarshal(b, &c); err != nil {
		t.Fatalf("parse %s: %v", dir, err)
	}
	return c
}

func hx(t *testing.T, s string) []byte {
	t.Helper()
	b, err := hex.DecodeString(s)
	if err != nil {
		t.Fatalf("bad hex: %v", err)
	}
	return b
}

// TestPerClassOctetExactRoundTrip: R-14.7 — every carriage class encodes to the oracle bytes and
// recovers its foreign message byte-identical (octet-exact), including non-canonical whitespace,
// proving the foreign bytes are never re-serialized (R-14.4). Byte equality with the shared
// oracle in Go and Rust ⟹ Go == Rust.
func TestPerClassOctetExactRoundTrip(t *testing.T) {
	for _, cd := range classDirs {
		c := loadClass(t, cd.Dir)
		if c.Class != cd.Class {
			t.Errorf("%s: class %d, want %d", cd.Dir, c.Class, cd.Class)
		}
		foreign := hx(t, c.ForeignHex)
		cb, err := carriage.Carry(c.ProtocolID, c.Class, c.ContentType, hx(t, c.CorrelationHex), c.Method, foreign)
		if err != nil {
			t.Fatalf("%s: carry: %v", cd.Dir, err)
		}
		if got := hex.EncodeToString(cb.Bytes()); got != c.BodyHex {
			t.Errorf("%s: body bytes\n got %s\nwant %s", cd.Dir, got, c.BodyHex)
		}
		// Round-trip: decode the carriage body and recover the foreign octets exactly.
		v, err := cbor.Decode(cb.Bytes())
		if err != nil {
			t.Fatalf("%s: decode: %v", cd.Dir, err)
		}
		rec, err := carriage.CarriageFromValue(v)
		if err != nil {
			t.Fatalf("%s: from value: %v", cd.Dir, err)
		}
		if !bytes.Equal(rec.Foreign, foreign) {
			t.Fatalf("%s: recovered foreign is not octet-identical", cd.Dir)
		}
		if rec.ProtocolID != c.ProtocolID || rec.Class != c.Class || rec.Method != c.Method {
			t.Errorf("%s: recovered fields differ", cd.Dir)
		}
	}
}

// TestOpaqueUndefinedProtocol: R-18.6 — an undefined protocol carries under OPAQUE on an
// experimental protocol id (no registration), byte-exact.
func TestOpaqueUndefinedProtocol(t *testing.T) {
	c := loadClass(t, "opaque")
	if carriage.ProtocolRange(c.ProtocolID) != "experimental" {
		t.Fatalf("opaque protocol id %#x not experimental", c.ProtocolID)
	}
	blob := []byte{0x00, 0x01, 0x02, 0xFF, 0xFE, 0x7F, 0x80}
	cb, err := carriage.Carry(c.ProtocolID, carriage.ClassOPAQUE, 1, nil, "", blob)
	if err != nil {
		t.Fatal(err)
	}
	v, _ := cbor.Decode(cb.Bytes())
	rec, err := carriage.CarriageFromValue(v)
	if err != nil || !bytes.Equal(rec.Foreign, blob) {
		t.Fatalf("opaque blob not recovered exactly: err=%v", err)
	}
}

// TestRegistry: the protocol registry CSV assignments are all in the standards range with the
// declared class, and the range boundaries classify per design §13.4.
func TestRegistry(t *testing.T) {
	f, err := os.Open(filepath.Clean("../../../vectors/registry/protocols.csv"))
	if err != nil {
		t.Fatal(err)
	}
	defer f.Close()
	rows, err := csv.NewReader(f).ReadAll()
	if err != nil {
		t.Fatal(err)
	}
	if len(rows) < 2 {
		t.Fatal("registry has no entries")
	}
	// header: protocol_id,name,class,range,reference
	for _, r := range rows[1:] {
		id, err := strconv.ParseUint(strings.TrimPrefix(r[0], "0x"), 16, 64)
		if err != nil {
			t.Fatalf("bad id %q: %v", r[0], err)
		}
		if carriage.ProtocolRange(id) != "standards" || r[3] != "standards" {
			t.Errorf("id %s: range mismatch (computed %s, csv %s)", r[0], carriage.ProtocolRange(id), r[3])
		}
	}
	// Range boundaries (design §13.4).
	for id, want := range map[uint64]string{0x00: "reserved", 0x01: "standards", 0x0F: "standards", 0x10: "experimental", 0x7F: "experimental", 0x80: "private", 0xFF: "private"} {
		if got := carriage.ProtocolRange(id); got != want {
			t.Errorf("ProtocolRange(%#x)=%s want %s", id, got, want)
		}
	}
}

// TestMappingErrorOnUnknownClass: R-14.8 — an unrepresentable class is a typed mapping error,
// never a silent drop.
func TestMappingErrorOnUnknownClass(t *testing.T) {
	if _, err := carriage.Carry(0x10, 99, 0, nil, "x", []byte("y")); err == nil {
		t.Fatal("unknown class carried")
	} else if ce, ok := err.(*cose.Error); !ok || ce.Kind != "MappingError" {
		t.Fatalf("want MappingError, got %v", err)
	}
}

// TestNotDelivered: R-14.8 — a below-foreign failure reports NotDelivered and never a false
// "delivered".
func TestNotDelivered(t *testing.T) {
	rep, err := carriage.Report(false)
	if err == nil || rep.Delivered {
		t.Fatal("failed delivery reported as delivered")
	}
	if ce, ok := err.(*cose.Error); !ok || ce.Kind != "NotDelivered" {
		t.Fatalf("want NotDelivered, got %v", err)
	}
	rep, err = carriage.Report(true)
	if err != nil || !rep.Delivered {
		t.Fatalf("successful delivery not reported: %v", err)
	}
}

// TestIdentityContainment: R-14.6 — a foreign principal named inside the foreign bytes confers no
// authority; the authorizing principal is the N-AALP signer of the carriage object. The carriage
// object is a normal signed envelope object (R-14.2) and its foreign body recovers octet-exact.
func TestIdentityContainment(t *testing.T) {
	var seed [mldsa65.SeedSize]byte
	for i := range seed {
		seed[i] = 80
	}
	pk, sk := mldsa65.NewKeyFromSeed(&seed)
	signer, verifier := cose.MLDSA65Signer{SK: sk}, cose.MLDSA65Verifier{PK: pk}
	kindOK := func(ch, k uint64) bool { return true }

	foreign := []byte(`{"jsonrpc":"2.0","method":"tools/call","params":{"from":"attacker-principal"}}`)
	cb, err := carriage.Carry(0x01, carriage.ClassJSONRPC, 0, []byte{1, 2, 3, 4}, "tools/call", foreign)
	if err != nil {
		t.Fatal(err)
	}
	obj := &envelope.Object{Kind: 0, Channel: 13, Tier: 0, Signer: pk.Bytes(), Created: 100, Effect: 0, Profile: uint64(cose.ProfilePublic), Body: cb.ToValue()}
	signed, err := envelope.Sign(obj, signer)
	if err != nil {
		t.Fatal(err)
	}
	o, err := envelope.Verify(cose.ProfilePublic, verifier, kindOK, nil, signed)
	if err != nil {
		t.Fatalf("verify: %v", err)
	}
	// The authority is the N-AALP signer, NOT the foreign principal.
	auth := carriage.CarriageAuthority(o)
	if !bytes.Equal(auth, pk.Bytes()) {
		t.Fatal("authority is not the N-AALP signer")
	}
	if bytes.Contains(auth, []byte("attacker-principal")) {
		t.Fatal("authority leaked from the foreign message")
	}
	// The foreign message still carries the (non-authoritative) principal, recovered octet-exact.
	rec, err := carriage.CarriageFromValue(o.Body)
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(rec.Foreign, foreign) {
		t.Fatal("foreign not recovered octet-exact from the signed object")
	}
	if !bytes.Contains(rec.Foreign, []byte("attacker-principal")) {
		t.Fatal("foreign principal should be present in the carried message (just not authoritative)")
	}
}

File: impl/php/test/carriage_test.php
<?php
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
//
// C12 foreign-carriage-by-class conformance for the PHP SDK (design.md §13; R-14.1..14.8, R-18.6),
// graded against the shared independent PER-CLASS corpus vectors/carriage/<class>/cases.json (NOT
// produced by this code). N-AALP carries a foreign agent protocol by wrapping its message,
// octet-for-octet, in a signed N-AALP carriage object whose effect/safety/identity/audit apply, and
// whose foreign body is interpreted by a carriage CLASS — not a bespoke per-protocol mapping (R-14.1).
// Five structured classes (JSONRPC, HTTP, MSG, STREAM, DOC) plus a universal OPAQUE class make any
// protocol carriable immediately on an experimental protocol id with no registration. The foreign field
// is carried VERBATIM and MUST NOT be re-serialized (R-14.4); the carriage object's signer remains the
// authority — a foreign principal never becomes an N-AALP authorization identity (R-14.6).
//
// CORPUS-GRADED (pure), per class subdir: the carriage body {1..6} byte-for-byte, and the round-trip
// identity — the foreign message recovers byte-identical (octet-exact) from the encoded body (R-14.7,
// R-14.4).
// BEHAVIOURAL (isolation): an unrepresentable class is a typed MappingError (never a silent drop,
// R-14.8); a below-foreign failure reports NotDelivered and never a false "delivered" (R-14.8); the
// protocol-id ranges classify per §13.4; an OPAQUE undefined protocol carries on an experimental id.
// ED25519-DEMONSTRATED (isolation, NOT corpus-graded): identity containment (R-14.6) — a real
// Ed25519-signed carriage object's authority is its N-AALP signer, NOT the foreign principal named
// inside the foreign bytes, which survive octet-exact through the full sign→serialize→decode cycle. PHP
// is PURE-ONLY for ML-DSA; the reference signs with ML-DSA on the Public floor, so the signature is
// demonstrated here with Ed25519 (envelope.signWithEd25519 + Cose signature verify).
//
// Written test-first: Naalp\Carriage is absent until Carriage.php lands, so this fails RED with a fatal
// "class not found"; truncating the recovered foreign in carriageFromValue flips the per-class
// "<class> foreign recovered octet-exact" round-trip checks.
//
// Run:  php -d extension=sodium -d extension=intl test/carriage_test.php   (from impl/php/)
require __DIR__ . '/../src/bootstrap.php';

use Naalp\Carriage;
use Naalp\CarriageBody;
use Naalp\NaalpObject;
use Naalp\Cbor;
use Naalp\Cose;
use Naalp\Envelope;
use Naalp\U;
use Naalp\B;

$fails = 0;
function check(string $name, string $got, string $want): void
{
    global $fails;
    if ($got === $want) {
        echo "  ok   $name\n";
    } else {
        $fails++;
        echo "  FAIL $name\n       got  $got\n       want $want\n";
    }
}

/** Run $fn and return the caught error's ->kind (or "no-error" / the class name). */
function err_kind(callable $fn): string
{
    try {
        $fn();
        return "no-error";
    } catch (\Throwable $e) {
        return \property_exists($e, 'kind') ? $e->kind : \get_class($e);
    }
}

/** Walk up to the repository's shared PER-CLASS corpus (the independent oracle) for a class dir. */
function carriage_vectors(string $dir): array
{
    $d = __DIR__;
    for ($i = 0; $i < 6; $i++) {
        $p = $d . "/vectors/carriage/$dir/cases.json";
        if (is_file($p)) {
            return json_decode(file_get_contents($p), true, 512, JSON_THROW_ON_ERROR);
        }
        $d = dirname($d);
    }
    throw new RuntimeException("vectors/carriage/$dir/cases.json not found");
}

echo "carriage conformance (PHP) — graded vs vectors/carriage/<class>/cases.json (per-class oracle)\n";

// 1. per-class byte parity + round-trip identity (octet-exact foreign) against each class's own oracle.
$classDirs = [
    "jsonrpc" => Carriage::CLASS_JSONRPC,
    "http" => Carriage::CLASS_HTTP,
    "msg" => Carriage::CLASS_MSG,
    "stream" => Carriage::CLASS_STREAM,
    "doc" => Carriage::CLASS_DOC,
    "opaque" => Carriage::CLASS_OPAQUE,
];
foreach ($classDirs as $dir => $wantClass) {
    $c = carriage_vectors($dir);
    check("$dir class code == oracle", (string) $c["class"], (string) $wantClass);
    $foreign = hex2bin($c["foreign_hex"]);
    $cb = Carriage::carry($c["protocol_id"], $c["class"], $c["content_type"], hex2bin($c["correlation_hex"]), $c["method"], $foreign);
    check("$dir carriage body == oracle", bin2hex($cb->bytes()), $c["body_hex"]);
    // round-trip: decode the carriage body and recover the foreign octets EXACTLY (R-14.7). This is the
    // load-bearing octet-for-octet property (R-14.4) — the mutation target.
    $rec = Carriage::carriageFromValue(Cbor::decode($cb->bytes()));
    check("$dir foreign recovered octet-exact", bin2hex($rec->foreign), $c["foreign_hex"]);
    check("$dir class recovered == oracle", (string) $rec->klass, (string) $c["class"]);
    check("$dir protocol_id recovered == oracle", (string) $rec->protocolId, (string) $c["protocol_id"]);
    check("$dir method recovered == oracle", $rec->method, $c["method"]);
}

// 2. R-18.6 — an OPAQUE undefined protocol carries on an EXPERIMENTAL protocol id (no registration),
//    byte-exact, and recovers an arbitrary binary blob unchanged.
$op = carriage_vectors("opaque");
check("opaque protocol id is experimental", Carriage::protocolRange($op["protocol_id"]), "experimental");
$blob = "\x00\x01\x02\xff\xfe\x7f\x80";
$cbOp = Carriage::carry($op["protocol_id"], Carriage::CLASS_OPAQUE, 1, "", "", $blob);
$recOp = Carriage::carriageFromValue(Cbor::decode($cbOp->bytes()));
check("opaque blob recovered octet-exact", bin2hex($recOp->foreign), bin2hex($blob));

// 3. §13.4 — the protocol-id ranges classify per the design boundaries.
$ranges = [0x00 => "reserved", 0x01 => "standards", 0x0F => "standards", 0x10 => "experimental", 0x7F => "experimental", 0x80 => "private", 0xFF => "private", 0x100 => "invalid"];
foreach ($ranges as $id => $want) {
    check("protocol range " . sprintf("0x%02x", $id), Carriage::protocolRange($id), $want);
}

// 4. R-14.8 — an unrepresentable class is a typed MappingError, never a silent drop.
check("unknown class rejected (MappingError)", err_kind(fn() => Carriage::carry(0x10, 99, 0, "", "x", "y")), "MappingError");
// and a body carrying an unknown class parses to a MappingError, not silently.
$badClassBody = (new CarriageBody(0x10, 99, 0, "", "x", "y"))->bytes();
check("unknown-class body rejected (MappingError)", err_kind(fn() => Carriage::carriageFromValue(Cbor::decode($badClassBody))), "MappingError");

// 5. R-14.8 — a below-foreign failure reports NotDelivered and never a false "delivered".
check("failed delivery reports NotDelivered", err_kind(fn() => Carriage::report(false)), "NotDelivered");
check("successful delivery reports delivered", Carriage::report(true)->delivered ? "true" : "false", "true");

// 6. the class-name vocabulary; an out-of-range code is "unknown".
foreach ($classDirs as $dir => $code) {
    check("class name of $code", strtolower(Carriage::className($code)), $dir);
}
check("unknown class name", Carriage::className(99), "unknown");

// 7. R-14.6 identity containment (ED25519-DEMONSTRATED, isolation): a real Ed25519-signed carriage
//    object's authority is its N-AALP SIGNER, never the foreign principal named inside the foreign
//    bytes, which survive octet-exact through the full sign→serialize→decode cycle.
$seed = str_repeat("\x50", 32);
$pk = sodium_crypto_sign_publickey(sodium_crypto_sign_seed_keypair($seed));
$foreignPk = sodium_crypto_sign_publickey(sodium_crypto_sign_seed_keypair(str_repeat("\x51", 32)));
$foreignMsg = '{"jsonrpc":"2.0","method":"tools/call","params":{"from":"attacker-principal"}}';
$cb = Carriage::carry(0x01, Carriage::CLASS_JSONRPC, 0, "\x01\x02\x03\x04", "tools/call", $foreignMsg);
$o = new NaalpObject(kind: 0, channel: 13, signer: $pk, created: 100, effect: 0, body: $cb->toValue(), profile: Cose::PROFILE_PUBLIC);
// the authority is the N-AALP signer, NOT the foreign principal.
check("carriage authority == N-AALP signer", bin2hex(Carriage::authority($o)), bin2hex($pk));
check("authority does not leak the foreign principal", str_contains(Carriage::authority($o), "attacker-principal") ? "leaked" : "clean", "clean");
// sign end-to-end with real Ed25519 and confirm the signature; a tampered signature and a foreign key fail.
$signed = Envelope::signWithEd25519($o, $seed);
[$prot, $payload, $sig] = Cose::parseSign1Raw($signed);
$tbs = Cose::toBeSignedRaw($prot, $payload);
check("signed carriage object verifies", Cose::ed25519Verify($pk, $tbs, $sig) ? "true" : "false", "true");
$badSig = $sig;
$badSig[0] = $badSig[0] ^ "\x01";
check("tampered carriage signature rejected", Cose::ed25519Verify($pk, $tbs, $badSig) ? "true" : "false", "false");
check("foreign key rejected", Cose::ed25519Verify($foreignPk, $tbs, $sig) ? "true" : "false", "false");
// recover the signer (field 5) and body (field 10) from the SIGNED bytes; the foreign survives octet-exact.
$bodyMap = Cbor::decode($payload);
$recSigner = null;
$recBody = null;
foreach ($bodyMap->pairs as [$k, $v]) {
    if ($k instanceof U && $k->v === Envelope::FIELD_SIGNER && $v instanceof B) {
        $recSigner = $v->v;
    }
    if ($k instanceof U && $k->v === Envelope::FIELD_BODY) {
        $recBody = $v;
    }
}
check("recovered signer is the authority (not the foreign principal)", bin2hex((string) $recSigner), bin2hex($pk));
$recCarriage = Carriage::carriageFromValue($recBody);
check("foreign survives octet-exact through the signed object", $recCarriage->foreign, $foreignMsg);
check("foreign still carries the (non-authoritative) principal", str_contains($recCarriage->foreign, "attacker-principal") ? "present" : "absent", "present");

echo ($fails === 0 ? "PASS" : "FAIL ($fails)") . "\n";
exit($fails === 0 ? 0 : 1);
